packages/gatsby/src/internal-plugins/query-runner/query-watcher.js

```diff
@@ -48,16 +48,22 @@ exports.extractQueries = () => {
 
 const runQueriesForComponent = componentPath => {
   const pages = getPagesForComponent(componentPath)
-  // Remove page data dependencies before re-running queries because
+  // Remove page & layout data dependencies before re-running queries because
   // the changing of the query could have changed the data dependencies.
   // Re-running the queries will add back data dependencies.
-  boundActionCreators.deletePagesDependencies(pages.map(p => p.path))
-  const component = store.getState().pageComponents[componentPath]
+  boundActionCreators.deleteComponentsDependencies(
+    pages.map(p => p.path || p.id)
+  )
+  const component = store.getState().components[componentPath]
   return Promise.all(pages.map(p => queryRunner(p, component)))
 }
 
-const getPagesForComponent = componentPath =>
-  store.getState().pages.filter(p => p.component === componentPath)
+const getPagesForComponent = componentPath => {
```

Comment:
Yea it's really awkward, I found naming in the query-runner very hard. I've tried to be as descriptive as possible thats why there are variable names like `pagesAndLayouts`, instead of just `components`.